apps/rush-lib/src/logic/TaskSelector.ts

```diff
@@ -60,143 +59,71 @@ export class TaskSelector {
   }
 
   public registerTasks(): TaskCollection {
-    if (this._options.toProjects.length > 0) {
-      this._registerToProjects(this._options.toProjects);
-    }
-    if (this._options.fromProjects.length > 0) {
-      this._registerFromProjects(this._options.fromProjects);
-    }
-    if (this._options.toProjects.length === 0 && this._options.fromProjects.length === 0) {
-      this._registerAll();
-    }
+    const selectedProjects: Set<RushConfigurationProject> = this._computeSelectedProjects();
 
-    return this._taskCollection;
+    return this._createTaskCollection(selectedProjects);
   }
 
-  private _registerToProjects(toProjects: ReadonlyArray<RushConfigurationProject>): void {
-    const dependencies: Map<string, RushConfigurationProject> = new Map<string, RushConfigurationProject>();
-
-    for (const toProject of toProjects) {
-      this._collectAllDependencies(toProject, dependencies);
-    }
-
-    // Register any dependencies it may have
-    for (const dependencyProject of dependencies.values()) {
-      this._registerTask(dependencyProject);
-    }
+  private _computeSelectedProjects(): Set<RushConfigurationProject> {
+    const { fromProjects, toProjects } = this._options;
 
-    if (!this._options.ignoreDependencyOrder) {
-      // Add ordering relationships for each dependency
-      for (const dependencyProject of dependencies.values()) {
-        this._taskCollection.addDependencies(
-          ProjectBuilder.getTaskName(dependencyProject),
-          dependencyProject.localDependencyProjects.map((x) => ProjectBuilder.getTaskName(x))
-        );
-      }
-    }
-  }
+    const hasFrom: boolean = !!(fromProjects && fromProjects.size);
+    const hasTo: boolean = !!(toProjects && toProjects.size);
 
-  private _registerFromProjects(fromProjects: ReadonlyArray<RushConfigurationProject>): void {
-    const dependentList: Map<string, Set<RushConfigurationProject>> = this._getDependentGraph();
-    const dependents: Map<string, RushConfigurationProject> = new Map<string, RushConfigurationProject>();
+    const expandedFrom: Set<RushConfigurationProject> | false =
+      hasFrom && expandAllDependents<RushConfigurationProject>(expandFromProjects(fromProjects));
 
-    for (const fromProject of fromProjects) {
-      this._collectAllDependents(dependentList, fromProject, dependents);
-    }
+    const expandedTo: Set<RushConfigurationProject> | false =
+      hasTo && expandAllDependencies<RushConfigurationProject>(expandToProjects(toProjects));
 
-    // Register all downstream dependents
-    for (const dependentProject of dependents.values()) {
-      this._registerTask(dependentProject);
+    if (expandedFrom && expandedTo) {
+      // Select all projects that depend on those in fromProjects OR that are dependencies of those in toProjects
+      // To include all dependencies of --from, per PR#2349, use expandAllDependencies(expandedFrom)
```

Comment:
The behavior desired in #2349 can be facilitated by use of either `expandAllDependencies(expandedFrom)` or applying a final round of `expandAllDependencies()` to the result.

I will still need a follow up means to call this class and request only the critical path between projects `X` (inclusive) and `Y` (exclusive) as that was my primary motivation for implementing this feature.

Specifically, the scenario is that a developer is watching (via `webpack-dev-server` or `jest` or whatever) project `Y`, and is making changes to some other project, e.g. `X` and wants to see the effects of their changes as quickly as possible in project `Y`.

Reply:
Added `--with-dependencies` per discussion.